File: writers/writeValidUtf8.js
#!/usr/bin/env node

// Read a binary file and write a utf-8 file.

function writeValidUtf8(inFilename, outFilename, skipOrReplace) {
  // Read the binary file input file, which might contain invalid
  // UTF-8 bytes, then write valid UTF-8 bytes to the output file
  // either skipping the invalid bytes or replacing them with U-FFFD.
  // When there is an error, display the error message to standard out
  // and return 1, else return 0.  The input file must be under 50k.

  const fs = require('fs')

  // Read the data as binary.
  try {
    var data = fs.readFileSync(inFilename)
  }
  catch (err) {
    console.log("Error reading the input file: " + inFilename)
    return 1
  }

  // Encode the data as UTF-8.
  var newData = Buffer.from(data, 'utf-8').toString();

  // Write the data.
  try {
    fs.writeFileSync(outFilename, newData)
  }
  catch (err) {
    console.log('Error writing the file: ' + outFilename)
    return 1
  }

  return 0 // success
}

function main() {
  // console.log(process.argv);

  // Parse the command line parameters.
  var msg = 'usage: writeValidUtf8t.js inFilename, outFilename, (skip | replace)'
  if (process.argv.length < 5) {
    console.log(msg)
    return 1
  }
  option = process.argv[4]
  if (option != 'skip' && option != 'replace') {
    console.log(msg)
    return 1
  }

  var inFilename = process.argv[2]
  var outFilename = process.argv[3]
  writeValidUtf8(inFilename, outFilename, option)
}

return main()

File: writers/writeValidUtf8.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Encode;

# Get the command line arguments.
my ($inFilename, $outFilename, $skipOrReplace) = @ARGV;
if (not defined $outFilename) {
  print "Usage writeValidUtf8 inFilename outFilename [skip|replace]\n";
  exit
}
if (not defined $skipOrReplace) {
 $skipOrReplace = "replace"
}
# print "$inFilename\n";
# print "$outFilename\n";
# print "$skipOrReplace\n";

# Read the input file into memory.
my $content;
open(my $fh, '<', $inFilename) or die "Unable to open the input file: '$inFilename'.\n";
{
  local $/;
  $content = <$fh>;
}
close($fh);
#print "$content\n";

# Decode the content as UTF-8.
my $utf8string;
if ($skipOrReplace eq "replace") {
  $utf8string = decode('UTF-8', $content);
} else {
  print "Skip currently not supported.\n";
  exit
  # Seems like you could use FB_QUIET to implement skip.
  # my($buffer, $string) = ("", "");
  # while (read($fh, $buffer, 256, length($buffer))) {
  #     $string .= decode($encoding, $buffer, Encode::FB_QUIET);
  #     # $buffer now contains the unprocessed partial character
  # }
}
#print "$utf8string\n";

# Write the UTF-8 string to the output file.
use open ':std', ':encoding(UTF-8)';
open(FH, '>', $outFilename) or die $!;
print FH $utf8string;
